### ISSUE TITLE
With 12.1 release async validators run when validatOn and asyncValidateOn are equal

### ISSUE THREAD

**Issue by greghawk** [davidkpiano/react-redux-form#839]
### The Problem
Sync validators that run on change and async validators that run on change call the async call on every keyup regardeless of if the synch validators are valid.  3rd party verification calls happen all the time now and they can be pricey.

### Steps to Reproduce
Make a text control with both sync and async validators.  set them both to fire on change.  They will fire on every keyup

### Expected Behavior
Async validators do not call until all sync validators pass. Preferably on keyup in the same cycle otherwise someone would have to type another character just to get async val to fire at the moment that the sync validators go valid.  I propose a new option for asyncValidateOn that would be afterSync that would effectively run the moment the sync validators go valid and then not fire on backspace or an invalid entry for sync validators is entered.

### Actual Behavior
Make a text control with both sync and async validators.  set them both to fire on change.  They will fire on every keyup

**Comment by davidkpiano** [comment]
What precludes you from having `asyncValidateOn="blur"` (which is the default), if it's too expensive? By having `asyncValidateOn="change"`, that means exactly that you want async validators to run on every single change.

**Comment by greghawk** [comment]
I hear what your saying but I have a wizard like form with a continue button. I am fighting the enter key to continue and the blur versus when someone clicks continue. If it validated right when the local validation was good it would both not call expensive validation calls (3rd party) when there was not valid data and be ready before they could hit enter or click continue. As it is now one has to blur then click and then click again. I could disable button but it looks hacky and a similar angular form I am trying to replace handles it with ease.

On Jun 15, 2017, at 5:47 PM, David Khourshid <notifications@github.com<mailto:notifications@github.com>> wrote:


What precludes you from having asyncValidateOn="blur" (which is the default), if it's too expensive? By having asyncValidateOn="change", that means exactly that you want async validators to run on every single change.

**Comment by leoalves** [comment]
You can manually call the async validator. After the change
```
changeAction={(model, value) => (dispatch) => {
  dispatch(actions.change(model, value))
  if(my value is valid) {
    dispatch(actions.asyncSetValidity(model, asyncValidator))
  }
}}
```

**Comment by greghawk** [comment]
I thought of trying that but I am unsure how to do that as I have component set to another component in order to support formatting which is slated as a coming enhancement.  I will attach my code file so you can see what I am getting at.

**Comment by greghawk** [comment]
Main control:
` <Control model={`formData.${step.name}`}
               component={formatter ? FormattedControl : undefined}
               {...formObj}
               validators={getValidations(step.validation)}
               asyncValidateOn="blur"
               asyncValidators={getAsyncValidations(step.validation)}
               validateOn="change"
               placeholder={step.placeholder}
               type={(step.mask) ? "password" : type}
               autoComplete={(step.mask) ? "new-password" : ""}
               autoFocus={steps.length === 1 || (steps.length > 1 && steps[0].name === step.name)}
               mapProps={{
                 className: ({fieldValue}) => (fieldValue.valid || fieldValue.pristine || !fieldValue.touched)
                   ? 'form-control'
                   : 'form-control has-error',
                 value: (props) => props.viewValue
               }}/>`

Component referenced as component = {FormattedControl} above

`<input
      type={props.type}
      autoComplete={props.autoComplete}
      autoFocus={props.autoFocus}
      className={props.className}
      placeholder={props.placeholder}
      onBlur={e => props.onBlur(props.modelParser(e.target.value))}
      value={props.formatter(props.value)}
      onKeyPress={e => props.onKeyPress(props.modelParser(e.target.value))}
      onFocus={e => props.onFocus(props.modelParser(e.target.value))}
      //onKeyDown={e => props.onKeyDown(props.modelParser(e.target.value))}
      onChange={e => props.onChange(props.modelParser(e.target.value))}
    />`

**Comment by leoalves** [comment]
Check if the value is valid before you call the async func.

```
<Control model={`formData.${step.name}`}
               component={formatter ? FormattedControl : undefined}
               {...formObj}
               validators={getValidations(step.validation)}
               asyncValidateOn="blur"
               asyncValidators={{
               isValueValid: (value, done) => {
               return (value is valid) ? getAsyncValidations(step.validation) : done(true)
               }
               }}
               validateOn="change"
               placeholder={step.placeholder}
               type={(step.mask) ? "password" : type}
               autoComplete={(step.mask) ? "new-password" : ""}
               autoFocus={steps.length === 1 || (steps.length > 1 && steps[0].name === step.name)}
               mapProps={{
                 className: ({fieldValue}) => (fieldValue.valid || fieldValue.pristine || !fieldValue.touched)
                   ? 'form-control'
                   : 'form-control has-error',
                 value: (props) => props.viewValue
               }}/>
```

Something like that. I didn't test but should work.

**Comment by greghawk** [comment]
@leoalves  does value is valid work as coded? I thought that was just the actual typed viewValue?

**Comment by greghawk** [comment]
By the way... Thanks so much for you guys help!!!!

**Comment by greghawk** [comment]
Love this platform

**Comment by greghawk** [comment]
@davidkpiano @leoalves Got this wot work but seems really nasty.
```
               changeAction={(model, value, {fieldValue}) => (dispatch, getState) => {
                 dispatch(actions.change(model, value));
                 if (step.validation.verification !== null && getState().forms.formData[step.name].valid) {
                   dispatch(actions.asyncSetValidity(`formData.${step.name}`, getAsyncValidations(step.validation).verification))
                 }
               }}
```

**Comment by greghawk** [comment]
@davidkpiano I got this to work except for the enter key issue.  It fires the rrf changeHandler in the source code and then loops my sync->async done loop again.  I tried various ways but cannot get ahold the the keypress in the changeaction nor can i prevent it inside of rrf on the enter key from happening.  This is the rather gruesome code I have mostly working.
```

               changeAction={(model, value, event) => (dispatch, getState) => {
                 dispatch(actions.change(model, value));
                 dispatch(actions.resetValidity(model, ["verification"]));
                 if (step.validation.verification !== null && getState().forms.formData[step.name].valid) {
                   dispatch(actions.setPending(model, true));
                   dispatch(actions.asyncSetValidity(model, (value, done) => getAsyncValidations(step.validation).verification(value).then(
                     (response) => {
                       dispatch(actions.setPending(model, false));
                       dispatch(actions.setTouched(model));
                       done({verification: response});
                     })));
                 }
               }}
               onKeyPress={e => {
                 if (e.key === 'Enter') {
                   e.preventDefault();
                   formActions.handleNextSteps();
                 }
               }}
```

**Comment by davidkpiano** [comment]
@greghawk Backtracking a little... would it solve your problem if async validation was always called on submit (whether via enter key or otherwise) and the form prevented from submitting until the async validators passed?

**Comment by greghawk** [comment]
@davidkpiano Sorry I did spam a bunch of info on this thread.  As a general thing for forms that may work but in my case I don't submit on enter.  I have a one or many questions at a time multi-step wizard type of form.  I only submit at the end of the wizard type process. There is a "continue" button on each page and then I accrue all the entered data in redux till I'm ready to submit to a backend.

**Comment by greghawk** [comment]
Also I hacked my way out of having the changeaction fire on enter key from rrf.  You'll laugh...

```
               onKeyPress={(e) => {
                 if (e.key === 'Enter') {
                   e.currentTarget.blur();
                   e.preventDefault();
                   e.key='PHENTER'  //LOL
                 }
               }}
```

**Comment by greghawk** [comment]
@davidkpiano So I am back in this loop again now.  I can't use the hack I have because if keypress happens quickly before onchange then when I validate I don't have async data yet.  What I am trying to do is have an async validation fire while the user is typing (after the data is valid).  I have tried this in the onchange handler but i end up with duplicate verification calls.

**Comment by davidkpiano** [comment]
@greghawk Are you on the latest version?

**Comment by greghawk** [comment]
I'm not as i had to do some interesting work you guys helped me with on the onchange handler to fire async only when valid and also make everything work on the enter key (native browser submit). I tried a few days ago with newer lib and had some issues

**Comment by davidkpiano** [comment]
Can you put those newer issues on to a new issue?

**Comment by greghawk** [comment]
@davidkpiano it is pretty much same issue but i can make a new one.  I tried 1.14.1 but when i set updateOn and asyncValidateOn to change then async runs on all keypress.  I really need and think others to would want it to run only after local validation is successful.  I can't do on blur as my product people want it to run after the field first goes valid.

**Comment by greghawk** [comment]
also the characters sent to async are one character off. they are missing the last typed character right when the field goes local valid. If you type a zipcode it does every char except 5th one but then does the asyc call when you type in 6 characters and sends them
